[TP] Refactor Parallel Style to make it more usable

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111176
* #111166
* __->__ #111160


One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```

fduwjj committed:
Refactor Parallel Style to make it more usable

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -334,8 +509,30 @@ class RowwiseParallel(ParallelStyle):
+            if _prepare_input is not None
+            else _get_prepare_input(
+                input_layouts,
+                [Shard(-1)] * len(input_layouts)  # type: ignore[arg-type]
```
I don't quite get this, if `isinstance(input_layouts, tuple)`, we should just use that directly? why need to duplicate the input layouts when generate the prepare_input_fn?

fduwjj replied:
No, because input_layouts can be not always [Shard(-1), Shard(-1)]?

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +367,140 @@ def make_output_reshard_tensor(
+    return make_output_shard_1d(output, device_mesh).to_local()  # type: ignore[call-arg, attr-defined, misc]
+
+
+def _get_prepare_input(
```
question about these functions: are we installing hooks by still using the prepare_input/prepare_output functions, that's why we need to generate prepare_input_output functions?

fduwjj replied:
that's the contract from the `distribute_module` API.

https://github.com/pytorch/pytorch/blob/8f90be447800a5c624d96b928387cb6f21509d37/torch/distributed/_tensor/api.py#L595

wanchaol replied:
I know, what I mean is that you are reusing the `_prepare_input_fn` to pass to distribute_module instead of generate a new function?

fduwjj replied:
right.

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -344,5 +541,27 @@ class ColwiseParallel(ParallelStyle):
+            if _prepare_input is not None
+            else _get_prepare_input(
+                input_layouts,
+                [Replicate()] * len(input_layouts)  # type: ignore[arg-type]
```
ditto: why we dupliacte the input layouts with Replicate() instead of reuse the input layouts when input layouts is tuple?

fduwjj replied:
What if input_layouts are [Shard(0), Shard(0)]?

wanchaol replied:
`[Shard(0), Shard(0)]` what does this mean? it should mean the first input arg is Shard(0) and the second is also Shard(0) right?

fduwjj replied:
For colwise parallel, input will always be converted to Replicate.

wanchaol reviewed (commented):
Some questions and suggestions, please see inline comments

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

fduwjj commented on torch/distributed/tensor/parallel/_utils.py:
```diff
@@ -12,6 +14,27 @@
     [DTensor, Optional[DeviceMesh], Optional[int]], Union[torch.Tensor, DTensor]
 ]
 
+LayoutsType = Union[Placement, Tuple[Placement, ...]]
```
Well, all the type alias in style.py are in this utils.

fduwjj commented on test/distributed/tensor/parallel/test_tp_examples.py:
```diff
@@ -64,7 +65,14 @@ def _test_mlp_training_e2e(self, is_seq_parallel=False, recompute_activation=Fal
             torch.arange(0, NUM_DEVICES),
         )
-        parallel_style = SequenceParallel() if is_seq_parallel else PairwiseParallel()
+        parallel_style = {
```
Removing will come later but now we will just add a warning here.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -57,15 +68,24 @@ class PairwiseParallel(ParallelStyle):
     """
-
-    def __init__(self, _prepare_input=None, _prepare_output=None) -> None:
+    @_deprecate_warnings("Use ColwiseParallel and RowwiseParallel instead.")  # type: ignore[misc]
```
I plan to remove the documentation in a another PR which will be out soon.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -82,11 +102,25 @@ class SequenceParallel(PairwiseParallel):
+        )
 
 
+@_deprecate_warnings("Specify input_layouts instead.")  # type: ignore[arg-type]
```
sure, will do in a separate PR.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -334,8 +509,30 @@ class RowwiseParallel(ParallelStyle):
-        super().__init__(_prepare_input, _prepare_output)
+    def __init__(
+        self,
+        input_layouts=Shard(-1),
```
I thought about this before, right now no one is setting `_prepare_input` and `_prepare_output` via position args?

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -37,9 +40,17 @@ class ParallelStyle(ABC):
     @abstractmethod
-    def __init__(self, _prepare_input, _prepare_output) -> None:
+    def __init__(
+        self, input_layouts, output_layouts, use_local, _prepare_input, _prepare_output
```
Not really, `_prepare_input` and `_prepare_output` are already used as kwargs. So the args can be ignored for now.

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -344,5 +559,33 @@ class ColwiseParallel(ParallelStyle):
+        *,
+        input_layouts=Replicate(),
+        output_layouts=Shard(-1),
+        use_local=True,
```
I feel everytime colwise parallel output would be consumed by a rowwise parallel style later (i.e. 3 col -> row, or just col-> row), maybe it's worth making the ColwiseParallel to return DTensor directly? wdyt?

fduwjj replied:
Not really, I think for MLP's case we can make the flag to be `False` because for Attn or some cases when DTensor has coverage issue, we always want to lower to tensor ops. If down the road, we are more confident about DTensor CPU overhead and op coverage, we can change this to False as default.

wanchaol reviewed (commented):
looks great! Have a few suggestions about some API details and would like to see if we can skip some cpu computation as we want to reduce CPU overhead

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -37,9 +40,23 @@ class ParallelStyle(ABC):
     _prepare_output: _PrepareOutputType
+    input_layouts: LayoutsType
+    output_layouts: LayoutsType
+    use_local: bool
```
If we want to make it clearer, I perfer `use_local_output `.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+                )
+            else:
+                return t
+        elif isinstance(t, DTensor):
```
sure.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+            else:
+                return t
+        elif isinstance(t, DTensor):
+            return t.redistribute(device_mesh, [output_layout])
```
I thought about this as well. Ideally this should be handled by inside redistribute rather than the caller. I can add this check for now.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+
+
+def _get_prepare_input(
+    input_layouts: LayoutsType, output_layouts: LayoutsType
```
Actually both layouts are required. Making it Optional will make implementation not very clean and we need to explain to users that they can skip redistribute which makes things complicated. I can add an early return in the redistribute.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+            dtensor = DTensor.from_local(
+                t, device_mesh, [input_layout], run_check=False
+            )
+            return dtensor.redistribute(device_mesh, [output_layout])
```
I prefer the check and will not take the Optional approach.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+            )
+
+        if not isinstance(input_layouts, tuple):
+            input_layouts = [  # type: ignore[assignment]
```
Well, it is consumed by the zip, will this make a difference in CPU overhead?

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+            else:
+                return t
+        else:  # t is DTensor.
+            dtensor = t.redistribute(device_mesh, [layout])
```
sure, I can add an early return check here.

fduwjj commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -325,7 +381,140 @@ def make_output_reshard_tensor(
+        Returns:
+            A :class:`DTensor` replicated over ``device_mesh``.
+        """
+        if not isinstance(outputs, tuple):
```
wait, this is the single check and we did early return already?

fduwjj replied:
I will add a comment to clarify a little bit.

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -324,8 +384,166 @@ def make_output_reshard_tensor(
+    return make_output_shard_1d(output, device_mesh).to_local()  # type: ignore[call-arg, attr-defined, misc]
+
+
+def _is_redistribute_necessary(
```
nit: can be fixed in follow up prs, name is not quite consistent with the check, i.e. if "==", it's actually not necessary, maybe call it like `_needs_redistribute`

fduwjj replied:
oh good point, let me update the func name.

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -344,5 +588,33 @@ class ColwiseParallel(ParallelStyle):
+        input_layouts=Replicate(),
+        output_layouts=Shard(-1),
+        use_local_output=True,
+    ) -> None:
```
are you also adding the documentations to `ColwiseParallel` and `RowwiseParalell` about what type of modules we supported, and the `input_layouts`, `output_layouts`, `use_local_output` documentation?

fduwjj replied:
They are added here: https://github.com/pytorch/pytorch/pull/111177

wanchaol reviewed (approved):
lgtm, we should probably add doc to colwise parallel + row parallel too

fduwjj commented:
@wanchaol docs are added in https://github.com/pytorch/pytorch/pull/111177.

fduwjj committed:
Update on "[TP] Refactor Parallel Style to make it more usable"



One thing we find it challenging for users is that we don't want to expose the concept of prepare_input and prepare_out to users since there are so many func names for users to select from which is quite confusing. On the other hand, the colwise and rowwise parallel always need input(out) and output(in) to be certain layout so we can somehow simplify the logic here and make it more usable.

So we added three public attributes to the parallelStyle here and the code logic is like:

```python
class ParallelStyle(ABC):
    """
    The parallel style user wants the module or submodule to be parallelized.
    We can add more in future, but this seems sufficient for immediate needs. Users can extend this class to build their own parallel style with customized input/output preparations.
  """
    input_layouts: Union[placement, Tuple[placement]]
    output_layouts: Union[placement, Tuple[placement]]
    use_local: bool
 
 
class RowwiseParallel(ParallelStyle):
    """
    Partitioning the row of a module. We assume the input to be a sharded DTensor and output to be a replicate Tensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(-1), output_layouts=Replicate(), use_local=True)


Class ColwiseParallel(ParallelStyle):
    """
    Partitioning the column of a module. We assume the input to be a Replicated DTensor and output to be a sharded DTensor.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(-1), use_local=True)

# For the case of Sequence parallel, users just set different input_shard, Shard(0) or Shard(1) instead of Replicate()

Class PrepareModuleInput(ParallelStyle):
    """
    Only used to specify the input distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Shard(0), output_layouts=Replicate(), use_local=False)

Class PrepareModuleOutput(ParallelStyle):
    """
    Only used to specify the output distribute spec for a module.
    """
    def __init__(self):
        super().__init__(input_layouts=Replicate(), output_layouts=Shard(0), use_local=True)

parallelize_plan = {
    "embedding": ColwiseParallel(output_shard=Replicate()),
    "attn": PrepareModuleInput(),
    "attn.w1": ColwiseParallel(),
    "attn.w2": ColwiseParallel(),
    "attn.w3": ColwiseParallel(),
    "attn.wo": RowwiseParallel(),
}

parallelize_module(
    module=block, # this can be a submodule or module
    device_mesh=mesh['tp'],
    parallelize_plan=parallelize_plan,
)
```



[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge -f "The failing test is not related to this PR and it succeeded in the following PR."
